ENH: Implement flattening for writer

PJBrs opened the pull request:
This is a vibe-coded attempt at implementing flattening of pdf forms. It is currently able to flatten text fields and button fields.

It is not at all integrated with the rest of the pypdf code. It introduces a dependency on pdfminer.six to calculate font widths. This should also be possible with some pypdf code but I don't think that pypdf has correct font metrics. (I tried with STANDARD_WIDTHS from
pypdf/_text_extraction/_layout_mode/_font.py , but text started overflowing certain fields.

So, basically a proof of principle.

Why this PR?

I think first and foremost to get some comments:

- I think that I can use lots of stuff from pypdf/constants.py.
- I also think that pypdf should have sufficient font metrics info and associated functions to be able to calculate font width, but I don't think that's the case now? I might be wrong though.
- I have some questions as well, such as, should flatten be a separate function, or should it be an argument to  update_page_form_field_values. That code, at least, has lots of functionality that could be reused.
- And I wonder what should go where. Some functions seem to be duplicating existing functionality.
- Others, I  think the wrap_text helper, are really new, but I wouldn't know what to call them.

I don't have a lot of python coding experience. Just for reference, I tested this code in a different project called dungeonsheets, see here for the correct branch: https://github.com/PJBrs/dungeon-sheets/tree/flatten_vibe; and here for the blank forms: https://github.com/canismarko/dungeon-sheets/tree/master/dungeonsheets/forms.

This PR is in response to https://github.com/py-pdf/pypdf/issues/232 .

stefan6419846 commented:
Thanks for the PR. Some notes:

* Parts of the formatting changes might be automated by running `ruff check --fix` locally.
* If possible, it is appreciated to re-use existing functionality of *pypdf*.
* If possible, please consider moving generic functionality out of the `PdfWriter` class.
* *pypdf* should not include another PDF library (*pdfminer.six*) for doing calculations. Instead, we should try to include the necessary code in *pypdf* directly. To avoid too large PRs, this might need to be part of a separate PR.

PJBrs commented:
@stefan6419846, thanks for your comments! I think what I need right now is a little bit more guidance. You wrote:
> Parts of the formatting changes might be automated by running ruff check --fix locally.

I only have python-3.9  and I don't know if ruff runs on that. But I can fix some of this manually as well.

> If possible, it is appreciated to re-use existing functionality of pypdf.

If you see any functionality that is duplicated, please tell me! From what I can tell:

- generate_appearance_stream does a lot of what I try to do with add_text_value
- I think it should be possible to integrate my flatten function with the existing update_page_form_field_values method
- The font width thing ought to be there already but it needs some improvement...

> If possible, please consider moving generic functionality out of the PdfWriter class.

If you can, please indicate what you consider generic functionality, and, more importantly, where it ought to go. I think my font_name_map function qualifies?

> pypdf should not include another PDF library (pdfminer.six) for doing calculations. Instead, we should try to include the necessary code in pypdf directly. To avoid too large PRs, this might need to be part of a separate PR.

I could try to correct STANDARD_WIDTHS from pypdf/_text_extraction/_layout_mode/_font.py . I don't know how to deal with encoding (yet)

Thanks for your comments!

stefan6419846 commented:
Please note that while I try to help you with getting these changes integrated, I do not know every aspect of the spec or implementation. My main goal is to ensure that any code fulfills our requirements, especially regarding maintenance.

Python 3.9 should not be an issue, as the latest published release (and version used by us) is still compatible to it.

Regarding re-use: You mentioned in your initial comment that integrating this into `update_page_form_field_values` would allow for less new code. This is what I have been referring to.

Some candidates for moving them out of the writer module:

* The font name map could go into the codecs module. It currently misses a source as well.
* `extract_formatting_from_annotation` and `add_*_field_value` could go into the annotations module.
* `base14_font_object` could go into the codecs or a font module, similar to `calculate_text_width` and `wrap_text`.
* `merge_content_streams` could in theory be part of the ContentStream-related classes.

> I could try to correct STANDARD_WIDTHS from pypdf/_text_extraction/_layout_mode/_font.py . I don't know how to deal with encoding (yet)

What do you mean by "correcting them"? Ideally, we would not use the functionality of the text extraction code here, but a dedicated implementation the text extraction code could use later on. This might be based upon the implementation from *pdfminer.six*, but we would have to analyze which parts of it we would actually have to port.

PJBrs commented:
> Please note that while I try to help you with getting these changes integrated, I do not know every aspect of the spec or implementation. My main goal is to ensure that any code fulfills our requirements, especially regarding maintenance.

Well, neither do I, but I do have the spec open to check some stuff :-) I've noted things that I didn't understand in comments.

In general, I've added most code to the `update_page_form_field_values` method and its helper function, deduplicated lots of code and removed the original `flatten` function, the `extract_formatting_from_annotation` function, and the `base14_font_object` function.

> Python 3.9 should not be an issue, as the latest published release (and version used by us) is still compatible to it.
> 
> Regarding re-use: You mentioned in your initial comment that integrating this into `update_page_form_field_values` would allow for less new code. This is what I have been referring to.
> 
> Some candidates for moving them out of the writer module:
> * `extract_formatting_from_annotation` and `add_*_field_value` could go into the annotations module.
OK, as it is, quite some of this functionality, as regards text fields, already was in the _writer code. I'd say that the `_update_field_annotation` that already existed is the equivalent of `add_button_field_value`

> * `base14_font_object` could go into the codecs or a font module, similar to `calculate_text_width` and `wrap_text`.
> * `merge_content_streams` could in theory be part of the ContentStream-related classes.

OK. I'll do that as a final step!
> > I could try to correct STANDARD_WIDTHS from pypdf/_text_extraction/_layout_mode/_font.py . I don't know how to deal with encoding (yet)
> 
> What do you mean by "correcting them"? Ideally, we would not use the functionality of the text extraction code here, but a dedicated implementation the text extraction code could use later on. This might be based upon the implementation from _pdfminer.six_, but we would have to analyze which parts of it we would actually have to port.

Well, I've tried reading character widths from the STANDARD_WIDTHS dict and suddenly text fields started overflowing that were OK when I still used pdfminer.six. Also, I think I checked that certain character widths were incorrect when compared to the Adobe afm files. For now, however, I'll switch to the default character width and see what happens.

I didn't understand much of how `generate_appearance_stream` works. I've added more comments and changed the offsets used in the text appearance stream generated there, otherwise all flattened fields would show up in the bottom-left corner. I don't see how that would disadvantage current use case, but if you could help me understand this code better, that would be great. I added my original appearance stream code as the `alt__appearance_stream` function, for reference.

PJBrs commented:
... Now I _do_ understand how `generate_appearance_stream` works, and I've found out how to correctly position XObjects without changing `generate_appearance_stream`. This PR is getting along fine :-)

PJBrs commented:
@stefan6419846  I think this PR now is in a state where it merits some more review: I deduplicated whatever I could and removed anything that wasn't immediately neccessary.

First, I added loads of comments to `_update_field_annotation` and `generate_appearance_stream` to make more sense of them. Are these comments OK or should I get rid of them in the final version of this PR?

For basic flattening, I needed to make some changes to `update_page_form_field_values` and `_update_field_annotation`, and to add the following methods:
- `merge_content_streams`, which merges a new content stream to existing page contents
- `add_apstream_object`, which prepares that content stream

Furthermore, the text flattening wouldn't make sense without being able to wrap text, so I changed `generate_appearance_stream` and added two other methods: `wrap_text` and `calculate_text_width`. I removed all font name and character matching for now, to first get this PR out of the way. Also, I removed the text alignment code for centering and right alignment, for now.

Some questions:

- I noticed that the text ends up being quite a bit smaller than previously, because `generate_appearance_stream` adds padding everywhere. I'm testing with a specific form, and I noticed that pdftk flattens it at quite a bit larger font size, and I'm quite certain that it works because it doesn't add 1 point of padding left, right, top and bottom. Is it OK to get rid of that?
- I have no idea why the tests/test_javascript.py::test_add_js, tests/test_javascript.py::test_added_js and tests/test_page_labels.py::test_index2label suddenly began to fail, seemingly without my code not getting near there, but I'll see what I can find.
- I still need to think of tests, that's probably going to be hard.

PJBrs commented:
I redid the PR, simplified it to three patches, and removed all bits involved in text wrapping to simplify this PR as much as possible.

The first two patches both add one method, together laying the groundwork for flattening. The third patch contains a small number of changes to update_page_form_field_values that enable flattening itself.

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,131 @@ def append_pages_from_reader(
+            font_res:The appearance stream's font resource (if given).
+        """
+        # Prepare XObject resource dictionary on the page
+        if "/XObject" not in cast(
```
Instead of doing `cast(DictionaryObject, page[PG.RESOURCES])` over and over, please prefer to store the typed value in a variable and re-use it.

PJBrs replied:
OK!

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,131 @@ def append_pages_from_reader(
+        # Sanitize the proposed object name to be a valid PDF name part (alphanumeric, underscore, hyphen)
+        # Replacing spaces with underscores, then removing any other non-alphanumeric/non-underscore/non-hyphen
+        sanitized_name = str(object_name).replace(" ", "_")
+        sanitized_name = re.sub(r"[^a-zA-Z0-9_-]", "_", sanitized_name)
```
We should have an utility function doing this somewhere instead of here.

PJBrs replied:
Do you mean there already is one? And if so, where is it? Or that I should write one? And if so, where should I add it?

PJBrs replied:
This works:

```
+    def _sanitize(self) -> "NameObject":
+        # Sanitize the proposed object name to be a valid PDF name part (alphanumeric, underscore, hyphen)
+        # Replacing spaces with underscores, then removing any other non-alphanumeric/non-underscore/non-hyphen
+        name = str(self)[1:]  # Remove leading forward slash
+        name = re.sub(r"\ ", "_", name)
+        name = re.sub(r"[^a-zA-Z0-9_-]", "_", name)
+        return NameObject("/" + name)
+
```

I only added one line between this and other methods, to keep with the style in the rest of the document.

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,131 @@ def append_pages_from_reader(
+        # Always add the resolved stream object to the writer to get a new IndirectObject.
+        # This ensures we have a valid IndirectObject managed by *this* writer.
+        xobject_ref = self._add_object(appearance_stream_obj)
+        # Create a name for the XObject. TODO Is this sufficiently unique?
```
What do you propose to do with this TODO?

PJBrs replied:
My proposal was to wait and see whether this holds up in practice, and only change this if problems pop up. As far as I understand, every annotation should have a specific name, so deriving a name from existing annotations should result in unique XObject names as well. Alternatively, a uuid could be added, but that seems overkill to me.

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,131 @@ def append_pages_from_reader(
+        # This matrix translates the XObject from its internal coordinate system
+        # (where its BBox is defined relative to its origin) to the target offset.
+        # TODO: This code does not deal with scaling.
+        a = 1
```
I propose to create a `Transformation` instance instead and implement a basic dumping method there.

PJBrs replied:
OK!

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,131 @@ def append_pages_from_reader(
+        f = y_offset
+        # Construct the PDF content stream commands to draw the XObject
+        xobject_drawing_commands = (
+            f"q\n{a:.4f} {b:.4f} {c:.4f} {d:.4f} {e:.4f} {f:.4f} cm\n{xobject_name} Do\nQ".encode()
```
Plain `.encode()` is discouraged as this might lead to different results depending on the environment this code is running in. Please specify an appropriate encoding.

PJBrs replied:
I'd expect this to be just ASCII compatible, this being just a series of floats and cm. Also, below, in generate_appearance_stream, none of the stream commands are encoded using a specific encoding, except for the text string itself. I thought it might be best to be consistent with what generate_appearance_stream does. Otherwise... ASCII coding?

stefan6419846 replied:
Thanks for the hint. Then I am okay with leaving the encoding out for now.

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -1090,35 +1218,42 @@ def update_page_form_field_values(
-                    )
-                else:
-                    parent_annotation[NameObject(FA.V)] = TextStringObject(value)
+                if not (value is None and flatten): # Only change values if given by user and not flattening.
```
```suggestion
                if not (value is None and flatten):  # Only change values if given by user and not flattening.
```

stefan6419846 replied:
How does this affect existing use cases?

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -1090,35 +1218,42 @@ def update_page_form_field_values(
                     annotation[NameObject(AA.AS)] = v
                     annotation[NameObject(FA.V)] = v
+                    if flatten and appearance_stream_obj is not None:
+                        # TODO: Should we add font resources to a flattened /Btn annotation?
```
What is about this TODO?

PJBrs replied:
For text field values, font information is collected in _update_field_annotation.  At this point in update_page_form_field_values, we haven't collected that info yet, so we can't pass it here either. An alternative would be to separate the font info collection in a separate method that we can call both here, in update_page_form_field_values, for button annotations, and in _update_field_annotation for text annotations.

I'm not sure whether this is necessary for button annotations, but it might be better to be safe rather than sorry. Let me know if you want me to add a patch to this effect.

PJBrs replied:
I'll try write one more patch to single out the font information collecting code. Better safe than sorry!

PJBrs replied:
On second thought, in the case of a /Btn annotation, we just copy the entire appearance stream, which is an XObject, which should have its own resources, including a font resource. I've adjusted the comment accordingly.

stefan6419846 commented on tests/test_writer.py:
```diff
@@ -1575,10 +1583,42 @@ def test_update_form_fields(tmp_path):
 
     Path(write_data_here).unlink()
 
+def test_merge_content_stream_to_empty_page():
```
Please ensure that there are two empty lines between two functions.

stefan6419846 reviewed (changes requested):
Thanks for your PR and patience.

I have added some inline comments. Additionally, I discovered that not all changes are really covered by the tests - although the code runs through, there is no check for the transformation matrix etc. in the code, which would ensure that everything is working correctly.

PJBrs commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,131 @@ def append_pages_from_reader(
+            if isinstance(existing_content, ArrayObject):
+                # Create a new StreamObject for the new_content_data
+                new_stream_obj = StreamObject()
+                new_stream_obj._data = new_content_data
```
OK!

PJBrs commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,131 @@ def append_pages_from_reader(
+                new_stream = StreamObject()
+                new_stream._data = merged_data
+                page[NameObject("/Contents")] = self._add_object(new_stream)
+        except KeyError:
```
OK!

PJBrs commented:
> Thanks for your PR and patience.
> 
> I have added some inline comments. Additionally, I discovered that not all changes are really covered by the tests - although the code runs through, there is no check for the transformation matrix etc. in the code, which would ensure that everything is working correctly.

Thanks for the review! I redid the PR with the changes you requested, excluding the non-resolved issues. I'll await your advice on those. Please check that I did the transformation thing right in the third patch.

PJBrs commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,121 @@ def append_pages_from_reader(
+        # Transformation matrix (a b c d e f) for 'a b c d e f cm'
+        xobject_transformation = Transformation()
+        xobject_translation = xobject_transformation.translate(x_offset, y_offset)
+        (a, b, _), (c, d, _), (e, f, _) = xobject_translation.matrix
```
I prepared an additional patch for this, which will be part of the next PR.

PJBrs replied:
Implemented as suggested.

PJBrs commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,121 @@ def append_pages_from_reader(
+            logger_warning(f"XObject '{xobject_name}' already added to page resources. This might be an issue.",
+                           __name__)
+        # Transformation matrix (a b c d e f) for 'a b c d e f cm'
+        xobject_transformation = Transformation()
```
Not applicable anymore due to below change!

PJBrs replied:
Actually still needed the above suggestion. Implemented.

PJBrs commented:
I redid the PR with all requested changes!

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,115 @@ def append_pages_from_reader(
+            font_name = font_res["/BaseFont"]  # [/"Name"] often also exists, but is deprecated
+            if "/Font" not in pg_res:
+                pg_res[NameObject("/Font")] = DictionaryObject()
+            if font_name not in cast(
```
To avoid casting twice, please use a corresponding variable to store the casting result in a variable and use it.

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,115 @@ def append_pages_from_reader(
+        # This ensures we have a valid IndirectObject managed by *this* writer.
+        xobject_ref = self._add_object(appearance_stream_obj)
+        # Create a name for the XObject. Note: Is this sufficiently unique?
+        # Sanitize the proposed object name to be a valid PDF name part (alphanumeric, underscore, hyphen)
```
The description of how sanitizing works should be part of the docstring implementing it, not of its usage.

PJBrs replied:
Fixed.

It took this to also mean that the new NameObject._sanitize method needed a docstring, so I changed the comment there into a docstring.

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,115 @@ def append_pages_from_reader(
+        else:
+            logger_warning(f"XObject '{xobject_name}' already added to page resources. This might be an issue.",
+                           __name__)
+        # Transformation matrix (a b c d e f) for 'a b c d e f cm'
```
This comment seems obsolete.

PJBrs replied:
Fixed

stefan6419846 commented on tests/test_writer.py:
```diff
@@ -1575,11 +1585,70 @@ def test_update_form_fields(tmp_path):
+    stream = contents_obj.get_object()
+    assert isinstance(stream, StreamObject)
+    expect = b"q\n1.0000 0.0000 0.0000 1.0000 200.0000 200.0000 cm\n/Fm_AA2342__________aa_-_ Do\nQ"
+    assert expect == stream._data
```
```suggestion
    assert stream.get_data() == expected
```

PJBrs replied:
Fixed, sorry for the oversight.

PJBrs commented:
@stefan6419846  

I redid the PR. It is not quite clear anymore to me when to use one line and when two lines distance between methods. I adhered to what was already there, and did two lines between tests and one line between methods. Also see the docstring in _sanitize, not entirely sure whether that was what you wanted.

I hope this good!

stefan6419846 commented:
> It is not quite clear anymore to me when to use one line and when two lines distance between methods.

The convention is to use two lines between functions, classes etc., while only one line is desired between methods.

stefan6419846 commented on pypdf/_writer.py:
```diff
@@ -865,12 +865,100 @@ def append_pages_from_reader(
+            object_name: The name of the appearance stream.
+            x_offset: The horizontal offset for the appearance stream.
+            y_offset: The vertical offset for the appearance stream.
+            font_res:The appearance stream's font resource (if given).
```
```suggestion
            font_res: The appearance stream's font resource (if given).
```

stefan6419846 reviewed (approved):
Thanks for your patience.
